src/utilities/nep_utilities.cuh

```diff
@@ -456,138 +498,100 @@ static __device__ __forceinline__ void get_f12_5body(
-  f12[0] += tmp * x;
-  f12[1] += tmp * y;
-  f12[2] += tmp * z;
-
-  // x
-  tmp = s[1] * z * (7.0f * z2 - 3.0f * r2 - 6.0f * x2);  // Y41_real
-  tmp += s[2] * (-6.0f * xyz);                           // Y41_imag
-  tmp += s[3] * 4.0f * x * (3.0f * z2 - x2);             // Y42_real
-  tmp += s[4] * 2.0f * y * (7.0f * z2 - r2 - 2.0f * x2); // Y42_imag
-  tmp += s[5] * 3.0f * z * x2my2;                        // Y43_real
-  tmp += s[6] * 6.0f * xyz;                              // Y43_imag
-  tmp += s[7] * 4.0f * x * (x2 - 3.0f * y2);             // Y44_real
-  tmp += s[8] * 4.0f * y * (3.0f * x2 - y2);             // Y44_imag
-  tmp *= 2.0f;
-  tmp += s[0] * 12.0f * x * (r2 - 5.0f * z2); // Y40
-  f12[0] += tmp * Fp * fn * 2.0f;
-  // y
-  tmp = s[1] * (-6.0f * xyz);                            // Y41_real
-  tmp += s[2] * z * (7.0f * z2 - 3.0f * r2 - 6.0f * y2); // Y41_imag
-  tmp += s[3] * 4.0f * y * (y2 - 3.0f * z2);             // Y42_real
-  tmp += s[4] * 2.0f * x * (7.0f * z2 - r2 - 2.0f * y2); // Y42_imag
-  tmp += s[5] * (-6.0f * xyz);                           // Y43_real
-  tmp += s[6] * 3.0f * z * x2my2;                        // Y43_imag
-  tmp += s[7] * 4.0f * y * (y2 - 3.0f * x2);             // Y44_real
-  tmp += s[8] * 4.0f * x * (x2 - 3.0f * y2);             // Y44_imag
-  tmp *= 2.0f;
-  tmp += s[0] * 12.0f * y * (r2 - 5.0f * z2); // Y40
-  f12[1] += tmp * Fp * fn * 2.0f;
-  // z
-  tmp = s[1] * 3.0f * x * (5.0f * z2 - r2);  // Y41_real
-  tmp += s[2] * 3.0f * y * (5.0f * z2 - r2); // Y41_imag
-  tmp += s[3] * 12.0f * z * x2my2;           // Y42_real
-  tmp += s[4] * 24.0f * xyz;                 // Y42_imag
-  tmp += s[5] * x * (x2 - 3.0f * y2);        // Y43_real
-  tmp += s[6] * y * (3.0f * x2 - y2);        // Y43_imag
-  tmp *= 2.0f;
-  tmp += s[0] * 16.0f * z * (5.0f * z2 - 3.0f * r2); // Y40
-  f12[2] += tmp * Fp * fn * 2.0f;
+  const float dx[3] = {(1.0f - r12[0] * r12[0]) * d12inv, -r12[0] * r12[1] * d12inv, -r12[0] * r12[2] * d12inv};
+  const float dy[3] = {-r12[0] * r12[1] * d12inv, (1.0f - r12[1] * r12[1]) * d12inv, -r12[1] * r12[2] * d12inv};
+  const float dz[3] = {-r12[0] * r12[2] * d12inv, -r12[1] * r12[2] * d12inv, (1.0f - r12[2] * r12[2]) * d12inv};
+
+  float z_pow[L + 1] = {1.0f};
+  for (int n = 1; n <= L; ++n) {
+    z_pow[n] = r12[2] * z_pow[n - 1];
+  }
+
+  float real_part = 1.0f;
+  float imag_part = 0.0f;
+  for (int n1 = 0; n1 <= L; ++n1) {
+    int n2_start = (L + n1) % 2 == 0 ? 0 : 1;
+    float z_factor = 0.0f;
+    float dz_factor = 0.0f;
+    for (int n2 = n2_start; n2 <= L - n1; n2 += 2) {
+      if (L == 1) {
+        z_factor += Z_COEFFICIENT_1[n1][n2] * z_pow[n2];
+        if (n2 > 0) {
+          dz_factor += Z_COEFFICIENT_1[n1][n2] * n2 * z_pow[n2 - 1];
+        }
+      }
+      if (L == 2) {
+        z_factor += Z_COEFFICIENT_2[n1][n2] * z_pow[n2];
+        if (n2 > 0) {
+          dz_factor += Z_COEFFICIENT_2[n1][n2] * n2 * z_pow[n2 - 1];
+        }
+      }
```

Comment:
Could one perhaps collect all the `Z_COEFFICIENTS` in a list or other structure, and just access them by `L`? Like `z_factor += Z_COEFFICIENTS[L][n1][n2]`? Would required less duplicate code.

Reply:
Using a 3D array will need 8*9*9 elements and here we are using `__constant__` in CUDA, which is very precious. It is better to keep the amount to a mininum.

Reply:
Ah right, got it